### PR TITLE
Added grpc monitoring with reflection api

### LINKED ISSUE #1117: Exec monitor
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

New Monitor

### 🔖 Feature description

Please add a monitor which executes user provided programs and checks the exit code.

### ✔️ Solution

The monitor executes program with optional arguments as provided by the user and checks the exit code. Users of the docker image would need to mount a directory with static binaries and shell scripts in order to use them.

e.g calling [gRPCurl](https://github.com/fullstorydev/grpcurl) to properly check if a gRPC services works. This is currently not possible and would mimic Kubernetes [exec probe](https://kubernetes.io/blog/2018/10/01/health-checking-grpc-servers-on-kubernetes/) or Monits [program status](https://mmonit.com/monit/documentation/monit.html#PROGRAM-STATUS-TEST) test.

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### PR DESCRIPTION
**Pull request by singhksandeep25**
# Description
- A pull request to add monitoring feature for gRPC services.
- Uses [grpcurl](https://github.com/fullstorydev/grpcurl) to make request to the grpc server
- It works by using grpc [reflection api](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md) on the server 
- If server is on, `grpcurl --plaintext localhost:50051 list` will list methods on the api, and will have a exit code of 0
- If server if down, `grpcurl --plaintext localhost:50051 list` will throw an error, and will have a exit code of 1

Fixes #1117

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

### CODE REVIEW

File: docker/alpine-base.dockerfile [hunk 1]
```diff
@@ -6,3 +6,8 @@ WORKDIR /app
 RUN apk add --no-cache iputils setpriv dumb-init python3 py3-cryptography py3-pip py3-six py3-yaml py3-click py3-markdown py3-requests py3-requests-oauthlib && \
     pip3 --no-cache-dir install apprise==0.9.8.3 && \
     rm -rf /root/.cache
+
+# Install grpcurl
+# Can be used like ./grpcurl
+RUN wget https://github.com/fullstorydev/grpcurl/releases/download/v1.8.6/grpcurl_1.8.6_linux_x86_64.tar.gz
+RUN tar xf grpcurl_1.8.6_linux_x86_64.tar.gz
```
Comment (Saibamen) [inline docker/alpine-base.dockerfile:13]: Can you create an argument variable to have version number in one place, instead of 3 places?

**Review by Saibamen — CHANGES_REQUESTED** [review]

File: docker/alpine-base.dockerfile (hunk 1 above)
Comment (singhksandeep25) [reply docker/alpine-base.dockerfile]: Done via adding ARG

**Comment by harryzcy** [comment]
@singhksandeep25 'replections' in pull request title is a typo, should be reflection.

File: docker/debian-base.dockerfile [hunk 2]
```diff
@@ -26,3 +26,8 @@ RUN node ./extra/download-cloudflared.js $TARGETPLATFORM && \
     rm -f cloudflared.deb && \
     apt --yes autoremove
 
+# Install grpcurl
+# Can be used like ./grpcurl
+ARG GRPCURL_VERSION=1.8.6
+RUN wget https://github.com/fullstorydev/grpcurl/releases/download/v$GRPCURL_VERSION/grpcurl_$GRPCURL_VERSION_linux_x86_64.tar.gz
+RUN tar xf grpcurl_$GRPCURL_VERSION_linux_x86_64.tar.gz
```
Comment (Computroniks) [inline docker/debian-base.dockerfile:33]: Will this support the docker image being used on arm based devices as well?

**Comment by louislam** [comment]
Unless there is no choice like Apprise or Cloudlfare Tunnels, please use Node.js library instead of external tool.

File: docker/debian-base.dockerfile (hunk 2 above)
Comment (singhksandeep25) [reply docker/debian-base.dockerfile:33]: Good catch, grpc_curl is available for arm 64 too, but not arm 7 which is also supported by the docker version of uptime kuma. Let me try to fix this.

**Comment by louislam** [comment]
#1964 is implemented with node.js library. I will likely choose #1964. Thanks for your contribution anyway.

**Comment by singhksandeep25** [comment]
Thanks @louislam . I too had a look at #1964  I personally couldn't find that particular node.js library.

**Comment by louislam** [comment]
Thanks for your contribution anyway, let's focus on https://github.com/louislam/uptime-kuma/pull/1964.

**Pull request closed without merging**